From 2bb0602556129b18536a44c86445bda2b0ba0a9f Mon Sep 17 00:00:00 2001
From: Peng Fan <peng.fan@nxp.com>
Date: Wed, 9 May 2018 21:15:16 +0800
Subject: [PATCH 3760/5242] MLK-18254 clk: imx: intpll: correct the
 programming flow

commit  fd2fe48243f1ec07a82d9268868019f2c986f9f2 from
https://source.codeaurora.org/external/imx/linux-imx.git

According to SPEC, when change the pll frequency and needs pll reset,
the t3 - t2 need to be greater than 1us and 1/FREF, respectively.
FREF is FIN / Prediv, the prediv is [1, 63], so choose
3us for safe.

The pll1443x does not have lock sel bit mask, so remove it.

Remove the bypass setting before changing frequency.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Reviewed-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/imx/clk-intpll.c |   25 +++++++++++++++++++------
 1 file changed, 19 insertions(+), 6 deletions(-)

diff --git a/drivers/clk/imx/clk-intpll.c b/drivers/clk/imx/clk-intpll.c
index e3ae0a6..a783a8a 100644
--- a/drivers/clk/imx/clk-intpll.c
+++ b/drivers/clk/imx/clk-intpll.c
@@ -11,6 +11,7 @@
 
 #include <linux/bitops.h>
 #include <linux/clk-provider.h>
+#include <linux/delay.h>
 #include <linux/err.h>
 #include <linux/io.h>
 #include <linux/slab.h>
@@ -197,7 +198,7 @@ static int clk_int_pll1416x_set_rate(struct clk_hw *hw, unsigned long drate,
 
 	/* Bypass clock and set lock to pll output lock */
 	tmp = readl_relaxed(pll->base);
-	tmp |= BYPASS_MASK | LOCK_SEL_MASK;
+	tmp |= LOCK_SEL_MASK;
 	writel_relaxed(tmp, pll->base);
 
 	/* Enable RST */
@@ -208,6 +209,14 @@ static int clk_int_pll1416x_set_rate(struct clk_hw *hw, unsigned long drate,
 		(rate->sdiv << SDIV_SHIFT);
 	writel_relaxed(div_val, pll->base + 0x4);
 
+	/*
+	 * According to SPEC, t3 - t2 need to be greater than
+	 * 1us and 1/FREF, respectively.
+	 * FREF is FIN / Prediv, the prediv is [1, 63], so choose
+	 * 3us.
+	 */
+	udelay(3);
+
 	/* Disable RST */
 	tmp |= RST_MASK;
 	writel_relaxed(tmp, pll->base);
@@ -250,12 +259,8 @@ static int clk_int_pll1443x_set_rate(struct clk_hw *hw, unsigned long drate,
 		return 0;
 	}
 
-	/* Bypass clock and set lock to pll output lock */
-	tmp = readl_relaxed(pll->base);
-	tmp |= BYPASS_MASK | LOCK_SEL_MASK;
-	writel_relaxed(tmp, pll->base);
-
 	/* Enable RST */
+	tmp = readl_relaxed(pll->base);
 	tmp &= ~RST_MASK;
 	writel_relaxed(tmp, pll->base);
 
@@ -264,6 +269,14 @@ static int clk_int_pll1443x_set_rate(struct clk_hw *hw, unsigned long drate,
 	writel_relaxed(div_val, pll->base + 0x4);
 	writel_relaxed(rate->kdiv << KDIV_SHIFT, pll->base + 0x8);
 
+	/*
+	 * According to SPEC, t3 - t2 need to be greater than
+	 * 1us and 1/FREF, respectively.
+	 * FREF is FIN / Prediv, the prediv is [1, 63], so choose
+	 * 3us.
+	 */
+	udelay(3);
+
 	/* Disable RST */
 	tmp |= RST_MASK;
 	writel_relaxed(tmp, pll->base);
-- 
1.7.9.5

